Makefile

```diff
@@ -46,8 +46,8 @@ check-style:
 	cd ${citus_abs_top_srcdir} && citus_indent --quiet --check
 .PHONY: reindent check-style
 
-# depend on install for now
-check: all install
+# depend on install-all so that downgrade scripts are installed as well
+check: all install-all
```

Comment:
I guess you mean the `check-*` targets in `src/test/regress` directory. They assume that you already compiled and installed the extension. They do not depend on any installation targets, and this is a good thing as you can test different scenarios quite easily.

Reply:
Got it, thanks.
Then we will not be able to fix `multi_extension` test failures that happens when user didn't install downgrade scripts, right ?

Reply:
a) `make check` on base directory will not fail. However `make install; make -C src/test/regress/ check-multi` will fail.

b) I can make all the tests that have multi_extension in their schedules depend on install-downgrades with the following patch:

```diff
diff --git a/src/test/regress/Makefile b/src/test/regress/Makefile
index 25a8d64be..d7d67ff27 100644
--- a/src/test/regress/Makefile
+++ b/src/test/regress/Makefile
@@ -120,7 +120,7 @@ check-empty: all
        $(pg_regress_multi_check) --load-extension=citus \
        -- $(MULTI_REGRESS_OPTS) $(EXTRA_TESTS)

-check-multi: all
+check-multi: all install-downgrades
        $(pg_regress_multi_check) --load-extension=citus \
        -- $(MULTI_REGRESS_OPTS) --schedule=$(citus_abs_srcdir)/multi_schedule $(EXTRA_TESTS)

@@ -132,7 +132,7 @@ check-multi-hyperscale-superuser: all
        $(pg_regress_multi_check) --conninfo="$(conninfo)" --worker-1-public-hostname="$(worker_1_public_hostname)" --worker-2-public-hostname="$(worker_2_public_hostname)" --load-extension=citus \
        -- $(MULTI_REGRESS_OPTS) --schedule=$(citus_abs_srcdir)/multi_schedule_hyperscale_superuser $(EXTRA_TESTS)

-check-multi-vg: all
+check-multi-vg: all install-downgrades
        $(pg_regress_multi_check) --load-extension=citus --valgrind \
        --pg_ctl-timeout=360 --connection-timeout=500000 --valgrind-path=valgrind --valgrind-log-file=$(VALGRIND_LOG_FILE) \
        -- $(MULTI_REGRESS_OPTS) --schedule=$(citus_abs_srcdir)/multi_schedule $(EXTRA_TESTS)
@@ -153,7 +153,7 @@ check-vanilla: all
 check-vanilla-internal:
        $(pg_regress_multi_check) --load-extension=citus --vanillatest

-check-multi-mx: all
+check-multi-mx: all install-downgrades
        $(pg_regress_multi_check) --load-extension=citus \
        -- $(MULTI_REGRESS_OPTS) --schedule=$(citus_abs_srcdir)/multi_mx_schedule $(EXTRA_TESTS)

@@ -215,3 +215,6 @@ check-citus-upgrade-mixed-local:
 clean distclean maintainer-clean:
        rm -f $(output_files) $(input_files)
        rm -rf tmp_check/
+
+install-downgrades:
+       $(MAKE) -C ../../backend/distributed/ install-downgrades
```

This change will ensure we install all the downgrades everytime.
```
❯ make --just-print -C src/test/regress check-multi
make: Entering directory '/home/hanefi/code/citus/src/test/regress'
make -C ../../backend/distributed/ install-downgrades
make[1]: Entering directory '/home/hanefi/code/citus/src/backend/distributed'
/usr/bin/install -c -m 644 /home/hanefi/code/citus/src/backend/distributed/build/sql/citus--10.0-2--10.0-1.sql /home/hanefi/code/citus/src/backend/distributed/build/sql/citus--9.2-4--9.2-2.sql /home/hanefi/code/citus/src/backend/distributed/build/sql/citus--9.4-1--9.3-2.sql /home/hanefi/code/citus/src/backend/distributed/build/sql/citus--9.5-1--9.4-1.sql /home/hanefi/code/citus/src/backend/distributed/build/sql/citus--10.1-1--10.0-2.sql /home/hanefi/code/citus/src/backend/distributed/build/sql/citus--10.0-1--9.5-1.sql /home/hanefi/code/citus/src/backend/distributed/build/sql/citus--9.3-2--9.2-4.sql '/home/hanefi/.pgenv/pgsql-13.2/share/extension/'
make[1]: Leaving directory '/home/hanefi/code/citus/src/backend/distributed'
'/usr/bin/perl' /home/hanefi/code/citus/src/test/regress/pg_regress_multi.pl --pgxsdir="/home/hanefi/.pgenv/pgsql-13.2/lib/pgxs" --bindir="/home/hanefi/.pgenv/pgsql-13.2/bin" --libdir="/home/hanefi/.pgenv/pgsql-13.2/lib" --majorversion="13" --postgres-builddir="/home/hanefi/.pgenv/src/postgresql-13.2"  --postgres-srcdir="/home/hanefi/.pgenv/src/postgresql-13.2" --load-extension=citus \
-- --inputdir=/home/hanefi/code/citus/src/test/regress   --launcher="/home/hanefi/code/citus/src/test/regress/log_test_times" --schedule=/home/hanefi/code/citus/src/test/regress/multi_schedule
make: Leaving directory '/home/hanefi/code/citus/src/test/regress'
```

I do not think this would be a bad thing. It may be overly verbose and unintuitive for some developers at the worst case.

Reply:
I guess the last changes you proposed is equal to = c82cf0a75eadee963a6fa3e1beabca274990329f ?

Tested with `check-multi` and this seems to be fixing `multi_extension` test failures, so I'm ok with merging this pr.

But let's have one of the seniors to take a look at too :)

Reply:
Also, not related with this pr but I guess we should also have `: all install-downgrades` in `check-columnar` too. Maybe we can fix it in this pr too ?

Reply:
> I guess the last changes you proposed is equal to = c82cf0a ?

No, c82cf0a is way more invasive than needs to be. I am not totally ok with using either of them.

> Also, not related with this pr but I guess we should also have : all install-downgrades in check-columnar too. Maybe we can fix it in this pr too ?

If the `columnar_schedule` contains any tests that actually depend on downgrade scripts, we may consider adding it to the patch. I do not want to increase test times for all schedules.

Reply:
>If the columnar_schedule contains any tests that actually depend on downgrade script

currently no

Reply:
>No, c82cf0a is way more invasive than needs to be. I am not totally ok with using either of them.

why this is the case, shouldn't we install downgrade scripts for each of the tests schedules ?

Reply:
> why this is the case, shouldn't we install downgrade scripts for each of the tests schedules ?
No, not really. If you do not run any `ALTER EXTENSION citus UPDATE TO <earlier version>` the downgrade scripts are not used.

Reply:
The patch I proposed above at https://github.com/citusdata/citus/pull/4717#discussion_r591417864 do not work. We should not attempt to install any files in regression makefile because some systems require sudo privileges and we would want to be able to run the tests as the unprivileged user.